python/llm/src/bigdl/llm/transformers/models/utils.py

```diff
@@ -119,3 +120,61 @@ def is_enough_kv_cache_room_4_31(past_key_value):
     # to determinate if is enough kv cache room in transformers between 4.31 and 4.35
     return past_key_value is not None and \
         past_key_value[0].stride()[1] > past_key_value[0].size(2) * past_key_value[0].size(3)
+
+
+def use_flash_attention(query):
+    if query.dim() == 3:
+        bsz, q_len, _ = query.size()
+    elif query.dim() == 4:
+        bsz, _, q_len, _ = query.size()
+    # check whether ipex flash attention can be used
+    if bsz > 1:
+        # only use flash attention for batch_size = 1 now
+        # as flash attention doesn't support attn_mask in ipex 2.1,
+        # so it will cause output error for padded batch input
+        return False
+    if q_len == 1:
+        # now only use flash attention for first token
+        # as it seems have no performance benifit for rest token now
+        return False
+    if query.device.type != "xpu":
+        # ipex flash attention only support for xpu
+        return False
+    ipex_version = get_ipex_version()
+    if ipex_version <= "2.0.110+xpu":
+        # ipex flash attention is supported from ipex 2.1
+        return False
+    if not torch.xpu.has_xetla():
+        # ipex flash attention is only supported for xetla
+        # may update this later
+        return False
+    if query.dtype not in [torch.float32, torch.float16]:
```

Comment:
How's `query.dtype` determined?

Reply:
It's determined by output from the previous layer .
In most cases, `query.dtype` is torch.float32.
Unless user pass `torch_dtype=fp16/bf16` or uses call `model.half().to('xpu')`, then here query.dtype will be fp16 or bf16.